src/main/java/fi/aalto/cs/apluscourses/utils/ArrayUtil.java

```diff
@@ -0,0 +1,39 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import java.util.function.IntFunction;
+import org.jetbrains.annotations.NotNull;
+
+public class ArrayUtil {
+
+  private ArrayUtil() {
+
+  }
+
+  /**
+   * Maps array to array.
+   *
+   * @param source    The source array.
+   * @param func      Mapping function that can throw an exception.
+   * @param generator Use {@code R[]::new}.
+   * @param <T>       Type of source elements.
+   * @param <R>       Type of result elements.
+   * @param <E>       Type of exception.
```

Comment:
the whole thing **is** about this `exception`, right?

Reply:
Not sure what you mean, but for functions that don't throw anything, we could also use
```java
Arrays.stream(source).map(func).toArray(generator)
```

We have quite a few of these kind of array mappings where array is populated in a `for` loop. Considering how `for` loops are a common source of bugs (especially off-by-1 errors) I think we should slowly start abstracting them from the business logic code.